core/src/main/java/io/lionweb/lioncore/java/model/impl/DynamicNode.java

```diff
@@ -76,8 +78,31 @@ && shallowClassifierInstanceEquality(parent, that.parent)
         && shallowClassifierInstanceEquality(concept, that.concept)
         && Objects.equals(propertyValues, that.propertyValues)
         && shallowContainmentsEquality(containmentValues, that.containmentValues)
-        && Objects.equals(referenceValues, that.referenceValues)
-        && Objects.equals(annotations, that.annotations);
+        && shallowReferenceEquality(referenceValues, that.referenceValues)
+        && shallowAnnotationsEquality(annotations, that.annotations);
+  }
+
+  private static boolean shallowReferenceEquality(
+      Map<String, List<ReferenceValue>> reference1, Map<String, List<ReferenceValue>> reference2) {
+    if (!reference1.keySet().equals(reference2.keySet())) {
+      return false;
+    }
+    return reference1.keySet().stream()
+        .allMatch(
+            referenceName -> {
+              List<ReferenceValue> references1 = reference1.get(referenceName);
+              List<ReferenceValue> references2 = reference2.get(referenceName);
+              return references1.size() == references2.size()
+                  && IntStream.range(0, references1.size())
+                      .allMatch(
+                          i ->
+                              Objects.equals(
+                                      references1.get(i).getResolveInfo(),
```

Comment:
Does `resolveInfo` *need* to be equal? I think only the `ID`s need to match.

Reply:
This is a good point. I am not sure if here we should check strict equality (but then we may want to do a deeper comparison for all aspects of nodes) or to consider "equivalence" instead. 

However I think resolveInfo may be important for this scenario: when we compare two references _not yet solved_ then these two references would both have the referred ID to null, while the resolveInfo may point to different things. In textual languages we may first parse code producing an AST (setting resolveInfo) and then resolve the references (consuming the resolving and set the referred ID). So there is a state when all references are not solved. If we looked only to the referred ID they would all be the same, while eventually they will be solved and point out to different elements

Reply:
We're talking instances of `ReferenceValue` here, right? So, we should anyway be able to compare the referred IDs and that's enough. The `resolveInfo`s are not guaranteed to be either not `null`, nor equal for equal reference targets. So, it should really only be a fallback: if refererred IDs are both `null`, then you can assume structural equality if the `resolveInfo`s are equal.

Reply:
Right, makes sense